packages/happy-dom/src/event/EventTarget.ts

```diff
@@ -47,10 +47,16 @@ export default abstract class EventTarget implements IEventTarget {
 	 */
 	public dispatchEvent(event: Event): boolean {
 		if (!event.target) {
-			event.target = this;
+			Object.defineProperty(event, "target", {
+				value: this,
+				writable: false
+			});
 		}
 
-		event.currentTarget = this;
+		Object.defineProperty(event, "currentTarget", {
```

Comment:
I am a bit worried about the performance impact of calling `defineProperty()` on each event and element where it bubbles. Using `defineProperty()` may change the shape of the object, which could take a way optimizations in the V8 engine.

Maybe we can do a solution where we use getters in the `Event` class which returns `_currentTarget` and `_target` that can be modified here?

We can also do some performance testing and see how it affects it.

Reply:
Hey 👋 ,

I gave your idea about the getters a quick shot and it did not work. The problem is that `jest-testing-library` uses `lib.dom.d.ts` where `target`/`currentTarget` are readonly. So as soon as we are writing directly to them and types are checked at runtime (`vitest`), this will fail :/

Tbh, I didn't even thought about performance in this context. I would be happy to do some microbenchmarks if there is some infrastructure I could use. Do you have anything at hand which could be repurposed?

Reply:
It is weird that `jest-testing-library` runs the compiler on the happy-dom internal code :thinking: 

Maybe it is possible to do something like this?

`Event.ts`
```javascript
export default class Event {
        ...
	public _currentTarget: IEventTarget = null;
	public _target: IEventTarget = null;

	/**
	 * Returns target.
	 *
	 * @returns Target.
	 */
	public get target(): string {
	      return this._target;
	}

	/**
	 * Returns current target.
	 *
	 * @returns Current target.
	 */
	public get currentTarget(): string {
	      return this._currentTarget;
	}
	
	...
}
```

`EventTarget.ts`
```javascript
export default class EventTarget {
        ...
	public dispatchEvent(event: Event): boolean {
		if (!event.target) {
			event._target = this;
		}

		event._currentTarget = this;

		....
	}
        ...
}
```

Reply:
Sadly this won't help as it's a typing issue. So whenever you try to write to `.target`/`.currentTarget`, it will fail because the types of `lib.dom.d.ts` mark those as readonly. This problem is independent of the implementation and the only fix (at least as far as I can see) is to not write directly to those readonly-props in the interfaces. That's also why the `defineProperty()`-approach works...

Reply:
@benbender I am not sure if you missed the "_" (underline) before the properties. My idea was that we perhaps can write to other properties instead and then return them from getters that are never changed? Then we will not risk any performance penalty of using `Object.defineProperty()`.

I would also like to understand why `vitest` or `jest-testing-library` are checking the types at runtime in the internal code of `happy-dom`. Do you have some example code I can run?

Reply:
Sure. Repro: https://github.com/benbender/happy-dom-repro-544

Reply:
Related: https://github.com/vitest-dev/vitest/issues/1494